File: Python/P0003.py
def is_prime(number: int) -> bool:
    return not (number & 1 == 0 or any(number % i == 0 for i in range(3, number, 2)))


def next_prime_factor(lower: int, upper: int) -> int:
    for number in range(lower, upper + 1):
        if is_prime(number) and upper % number == 0:
            return number
    return upper


def largest_prime_factor(upper: int, lower=None) -> int:
    if lower is None:
        lower = 2
    npf = next_prime_factor(lower, upper)
    if npf == upper:
        return upper
    else:
        return largest_prime_factor(upper // npf, npf)


def main() -> None:
    assert largest_prime_factor(13195) == 29
    answer = largest_prime_factor(600851475143)
    print(f"Answer: {answer}")


if __name__ == "__main__":
    main()

File: Go/P0003.go
package main

import (
	"fmt"
	"math"
)

func is_prime(number int) bool {
	if number%2 == 0 {
		return false
	}
	for num := 3; num < int(math.Sqrt(float64(number))); num++ {
		if number%num == 0 {
			return false
		}
	}
	return true
}

func find_prime_factors(number int) []int {
	var prime_factors []int
	for num := 2; num < int(math.Sqrt(float64(number))); num++ {
		if is_prime(num) && number%num == 0 {
			prime_factors = append(prime_factors, num)
		}
	}
	return prime_factors
}

func find_max(numbers []int) int {
	var max int = 0
	for _, number := range numbers {
		if number > max {
			max = number
		}
	}
	return max
}

func main() {
	var answer int = find_max(find_prime_factors(600851475143))
	fmt.Printf("Answer: %v\n", answer)
}
